lib/msal-angular/src/msal.interceptor.ts

```diff
@@ -65,26 +67,33 @@ export class MsalInterceptor implements HttpInterceptor {
         this.authService.getLogger().infoPii(`Interceptor - [${scopes}] scopes found for ${req.url}`);
 
         // Note: For MSA accounts, include openid scope when calling acquireTokenSilent to return idToken
-        return this.authService.acquireTokenSilent({...authRequest, scopes, account })
+        return this.msalBroadcastService.inProgress$
             .pipe(
-                catchError(() => {
-                    this.authService.getLogger().error("Interceptor - acquireTokenSilent rejected with error. Invoking interaction to resolve.");
-                    return this.acquireTokenInteractively(authRequest, scopes);
-                }),
-                switchMap((result: AuthenticationResult)  => {
-                    if (!result.accessToken) {
-                        this.authService.getLogger().error("Interceptor - acquireTokenSilent resolved with null access token. Known issue with B2C tenants, invoking interaction to resolve.");
-                        return this.acquireTokenInteractively(authRequest, scopes);
-                    }
-                    return of(result);
-                }),
-                switchMap((result: AuthenticationResult) => {
-                    this.authService.getLogger().verbose("Interceptor - setting authorization headers");
-                    const headers = req.headers
-                        .set("Authorization", `Bearer ${result.accessToken}`);
-
-                    const requestClone = req.clone({headers});
-                    return next.handle(requestClone);
+                filter((status: InteractionStatus) => status === InteractionStatus.None),
+                take(1),
+                switchMap(() => {
+                    return this.authService.acquireTokenSilent({...authRequest, scopes, account })
+                        .pipe(
+                            catchError(() => {
+                                this.authService.getLogger().error("Interceptor - acquireTokenSilent rejected with error. Invoking interaction to resolve.");
+                                return this.acquireTokenInteractively(authRequest, scopes);
```

Comment:
What happens if interaction was not in progress when `acquireTokenSilent` was invoked but then it is in progress when it reaches here?

Ideally I'd like to see the flow look something like this:
1. Invoke `acquireTokenSilent` regardless of interaction status
2. When ATS throws, check interaction status
3. If interaction is not currently in progress invoke interaction
4. If interaction is in progress, wait until it is done then go back to step 1

Reply:
@tnorling made some changes. Now it follows your logic

Reply:
Looks good, thanks! Any tests you can add?